### ISSUE TITLE
Tensorflow-GPU costing 3130% cpu resouces not for computation

### ISSUE THREAD

**Issue by ghostplant** [tensorflow/tensorflow#19073]
My host has 16 core cpu (with x2 hyper-thread), when I run resnet50 (weights & sync on gpu) benchmark for Tensorflow CUDA version, it actually not need CPU to do any heavy job since training and weights merge are done at gpu device. But it costs nearly full CPU resources not for computing but for long pulling queries on `cuEventQuery`, and 99% of `cuEventQuery` replies NOT_FINISHED, this is extremely bad for cpu utilization, and block all CPU cores from doing other tasks.

According to my profiling, for Resnet50, Tensorflow costs only 14840 time of model forward and backward, but it gathers all threads to do full-time queries, and should cost 23564854 times to do useless `cuEventQuery` due to not ready, which is incredibly heavy and triggers 3130% CPU cost on Tensorflow GPU computational version.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by ghostplant** [comment]
Have I written custom code: No
OS Platform and Distribution: Ubuntu 16.04 (64bit)
TensorFlow installed from: pip3 install tensorflow-gpu==1.4.2
TensorFlow version: 1.4.2
Bazel version: pre-built from pip
CUDA/cuDNN version: CUDA8 + CUDNN6
GPU model and memory: Tesla P100 (x4GPU, each 16GMem) + NVlink
Exact command to reproduce: `./tf_cnn_benchmarks/tf_cnn_benchmarks.py --model resnet50 --batch_size 64 --num_batches 50000 --data_format=NCHW --local_parameter_device gpu --data_dir=/usr/imagenet`
Source scripts: https://github.com/tensorflow/benchmarks/tree/master/scripts/tf_cnn_benchmarks

**Comment by karmel** [comment]
@reedwm , thoughts on the proper parameterization here?

**Comment by reedwm** [comment]
I cannot really reproduce. Adding a log on [this line](https://github.com/tensorflow/tensorflow/blob/714f3c4f2f901e865bfcbf830485adafb92dca48/tensorflow/stream_executor/cuda/cuda_driver.cc#L1053) and running for five steps, cuEventQuery was run only 10,000 times. When you said "and should cost 23564854 times to do useless cuEventQuery due to not ready", do you mean cuEventQuery was run 23564854 times? Was that over 50,000 steps?

Note that the CPU should be doing a fair amount of work for the input pipeline. You can run without the input pipeline by not specifying `--data_dir`. Do you still have the problem even when you do not specify `--data_dir`?

**Comment by ghostplant** [comment]
@reedwm Yes, if I do not specify --data-dir, the cpu occupation can be reduced, but it is meaningless when training a dataset. If I reduce the inter-thread parallel number, the cpu can be reduced as well but the model benchmark will be very bad.

**Comment by ghostplant** [comment]
The number of cuEventQuery depends on the GPU performance, if GPU is normal, the GPU function execution will be slow, which will highly increase the the amount of cuEventQuery.

**Comment by reedwm** [comment]
I am still not convinced the high CPU utilization is due to cuEventQuery. The CPU should be high utilized due to the input pipeline that is used when `--data_dir` is set. As you said, training is meaningless without `--data_dir`, but for real training, the CPU needs to do a lot of work for the input pipeline and so will have high utilization. Do you have any evidence that the CPU has high utilization because of cuEventQuery?

I'm not sure what you mean by you last comment: "The number of cuEventQuery depends on the GPU performance, if GPU is normal, the GPU function execution will be slow, which will highly increase the the amount of cuEventQuery." Can you clarify?

Also, @yzhwang, do you have any thoughts/comments?

**Comment by ghostplant** [comment]
I just profiled the reason, and here is my explanation:

By default, the `num_inter_threads` is 0, and my machine has 32 cores, and I found that `num_inter_threads` actually selects 32 by tensorflow's recommendation. Thus during the whole benchmark of tf_cnn_benchmark, I found 41 threads created by tensorflow in total: 1 thread for boot, 8 thread for num_intra_threads, and 32 thread for num_inter_threads.
Here we only consider 32 thread for num_inter_threads which is too heavy, and what I found is that every thread within these 32 num_inter_threads are occupied with cuEventQuery calls, which seems like everyone is eager to feed data to a available cuda stream as much as possible but every thread are doing `cuEventQuery` all the time, so this makes quantities of cpus are wasted on each thread calling `cuEventQuery`.

This issue can be slightly reduced by manual providing num_inter_threads a small number and the default tensorflow recommendation is not well which triggers ~3200% cpus occupation.

**Comment by ghostplant** [comment]
By using sympathy data source, the cuEventQuery is very lite, and CPU occupation is ~400%.

After giving `--data-dir`, it actually just increase ~10 additional `cuMemcpyHtoDAsync` calls to CUDA which reaches 3200% cpu occupation and this is definitely not a expected thing.

**Comment by ghostplant** [comment]
By using synthetic data source, here is the call counts:
```sh
...
cuEventQuery  x 23957788 times
...
cuMemcpyHtoDAsync_v2 x 590 times
cuMemcpyDtoHAsync_v2 x 60 times
cuMemcpyDtoDAsync_v2 x 63300 times
...
cudnnBatchNormalizationBackward x 12720
...
```
It only costs ~400% cpus which is quick;

However, by using real `--data-dir`, here is the call counts:
```sh
...
cuEventQuery  x 22843932 times
...
cuMemcpyHtoDAsync_v2 x 830 times
cuMemcpyDtoHAsync_v2 x 60 times
cuMemcpyDtoDAsync_v2 x 63300 times
...
cudnnBatchNormalizationBackward x 12720
...
```
It displays that 99% CUDA calls are the same for above 2 cases expect using `--data-dir` will have 240 more `cuMemcpyHtoDAsync_v2` (extactly all 240 times of  NCHW batch inputs for every mini-batch), but the CPU occupation increased to ~3200% CPUs. Can you explain what all these CPUs are doing in this case?

**Comment by mrry** [comment]
As @reedwm already pointed out, the higher CPU utilization arises because there is more work for the CPU to do per batch, including file I/O, protocol buffer parsing, JPEG decoding, distortions, and so on.

I'm not sure what your basis is for the statement "it still not necessary to cost so much CPU resources". Almost all of the low hanging fruit has been picked, although you might want to upgrade from TensorFlow 1.4.2 to TensorFlow 1.8, as the team has improved many of the operations in the input pipeline, and I'd expect CPU utilization to decrease with an upgrade. Nevertheless, the baseline for utilization will inherently be higher than the synthetic data case.

**Comment by ghostplant** [comment]
Thanks, I'll try on version 1.8.

**Comment by ghostplant** [comment]
TF1.8 is actually even slower than TF1.4 to process tensorflow/benchmark:tf_cnn_bencmark.py.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @reedwm: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ghostplant** [comment]
Thanks, all. I will close this PR. I think TF using CPU to do image preprocessing is limited by slow python runtime, and using an independent C-based preprocessor instead of TF runtime preprocessing solves my problem, not only faster but also less CPU occupation.
